tests/player.c

```diff
@@ -536,6 +439,451 @@ START_TEST (test_play_audio_eos)
+  new_stream = (GstPlayerStreamInfo *)
+                gst_player_get_current_audio_track (args->player);
+
+  fail_unless_equals_int (
+                gst_player_stream_info_get_index (new_stream),
+                args->index);
+
+  args->new_state->test_data = GINT_TO_POINTER (step + 1);
+  return FALSE;
+}
+
+static gboolean
+switch_subtitle_stream (gpointer user_data)
+{
+  GstPlayerStreamInfo *new_stream;
+  SwitchStreamArgs *args = (SwitchStreamArgs *) user_data;
+  gint step = GPOINTER_TO_INT (args->new_state->test_data);
+
+  fail_unless (gst_player_set_subtitle_track
+                (args->player, args->index) == TRUE);
+  new_stream = (GstPlayerStreamInfo *)
+                gst_player_get_current_subtitle_track (args->player);
+  fail_unless_equals_int (
+                gst_player_stream_info_get_index (new_stream),
+                args->index);
+
+  args->new_state->test_data = GINT_TO_POINTER (step + 1);
+  return FALSE;
+}
+
+static void
+switch_streams (GstPlayer * player, GList * list, TestPlayerState * new_state)
+{
+  gint i = 1;
+
+  /* go through the list of available stream index and switch
+   * to non active stream index.
+   */
+  for (; list != NULL; list = list->next) {
+    SwitchStreamArgs *args = g_new0 (SwitchStreamArgs, 1);
+    GstPlayerStreamInfo *stream = (GstPlayerStreamInfo *) list->data;
+
+    args->player = player;
+    args->index = gst_player_stream_info_get_index (stream);
+    args->new_state = new_state;
+
+    if (!strcmp (gst_player_stream_info_get_stream_type (stream), "audio"))
+      g_timeout_add_full (G_PRIORITY_DEFAULT, 3000 * i, switch_audio_stream, args,
+          (GDestroyNotify) g_free);
+    else if (!strcmp (gst_player_stream_info_get_stream_type (stream), "subtitle"))
+      g_timeout_add_full (G_PRIORITY_DEFAULT, 3000 * i, switch_subtitle_stream, args,
+          (GDestroyNotify) g_free);
+    i++;
+  }
+}
+
+static gboolean
+switch_streams_completed_loop (gpointer user_data)
+{
+  TestPlayerState *state = (TestPlayerState *) user_data;
+  gint step = GPOINTER_TO_INT (state->test_data);
+
+  if (step != 10)
+    return TRUE;
+
+  /* if all the stream switching is completed then quit the loop */
+  g_main_loop_quit (state->loop);
+  return FALSE;
+}
+
+static void
+test_play_stream_selection_cb (GstPlayer * player, TestPlayerStateChange change,
+    TestPlayerState * old_state, TestPlayerState * new_state)
+{
+  gint running = GPOINTER_TO_INT (new_state->test_data);
+
+  /* If pipeline is playing and stream switching thread is not started
+   * then switch to next available streams */
+  if (new_state->state == GST_PLAYER_STATE_PLAYING && !running) {
+    GstPlayerMediaInfo *media_info;
+
+    media_info = gst_player_get_media_info (player);
+    switch_streams (player, gst_player_get_subtitle_streams (media_info), new_state);
+    switch_streams (player, gst_player_get_audio_streams (media_info), new_state);
+    g_idle_add (switch_streams_completed_loop, new_state);
```

Comment:
Actually in callback, I am using loop and test_data variables and based on the condition will either increment the step or quit the loop.  I believe these variables are not updated in player cb. But if recommendation is to create a copy then sure I can do that. let me know your thought. thanks